### PR TITLE
[CHORE, TESTING] Test code can run form the front-end in all languages

### LINKED ISSUE #3760: [CYPRESS] Test run button in non-English too
In https://github.com/hedyorg/hedy/pull/3754 we test the run button (great!) but it would be nice if we also try setting the UI to a few non-English languages: Dutch, Hindi, Arabic for example (latin l2r, non-latin l2f, non-latin r2l)

### PR DESCRIPTION
**Pull request by TiBiBa**
Verify that the code can be ran in all languages:

This PR fixes #3760

### CODE REVIEW

**Comment by Felienne** [comment]
Hi @TiBiBa!

I think we said in the meeting that we'd close this and do the testing of the run button in a "manual" way, right (by repeatedly 'clicking' a language and then the run button)

**Comment by martinj2001** [comment]
@TiBiBa I think I fixed your problems with this tests, let me know if there is more to fix!

File: tests/cypress/e2e/hedy_page/editor_box.cy.js [hunk 1]
```diff
@@ -3,18 +3,12 @@ import {goToHedyPage} from "../tools/navigation/nav";
...
-          cy.get(element).should('have.text', 'print hello world!');
-        }
-        if(index == 1) {
-          cy.get(element).should('have.text', 'ask What is your name');
-        }
-        if(index == 2) {
-          cy.get(element).should('have.text', 'echo hello');
-        }
-      })
+      // click on textaread to get focus
```
Comment (Felienne) [inline tests/cypress/e2e/hedy_page/editor_box.cy.js:6]: I think this works nicely! But I think it only tests with English language setting. It would be nice to also test it for a few other languages, can that be done?

Comment (Felienne) [inline tests/cypress/e2e/hedy_page/editor_box.cy.js:6]: I think this works nicely! But I think it only tests with English language setting. It would be nice to also test it for a few other languages, can that be done? (Code in English should work in any language)

**Comment by jpelay** [comment]
Hello everyone! I'm turning this into a draft while we keep working on it!

**Comment by Felienne** [comment]
Maybe @TiBiBa can take a look? 🙏

**Review by Felienne — CHANGES_REQUESTED** [review]
Hi @jpelay 

I tried to test the tests as follows, but it seems to still work?

1. Break a print keyword, f.e. in French:

`_PRINT: ("affiche" | "prt") _SPACE?`

2. Be sure to not rerun on-deploy or it will restore the grammar!

**Comment by jpelay** [comment]
Hi @Felienne! Thank you for pointing it out!

It does seem like the page is refreshing, but the language is not changing? This one was supposed to show content from Bulgarian (notice that `switch-language-bg` was selected, but instead is still showing content from Arabic. It does appear that after changing the language once, that one keeps locked!

image

**Comment by Felienne** [comment]
Hi @jpelay!!

I might have a hunch as to why this does not test all languages!!! (@rix0rrr gave me the idea)

The current language is not present in the dropdown, so if you start with the first one (Arabic) will be selected:

image

Then we pick the second one, Bulgarian:

image

Now... the next one is Bengali:

image

This might mean we are skipping languages (I am not sure) So we first need to populate a stable list of languages and iterate over these.

**Comment by Felienne** [comment]
> It might be, but I'm actually iterating trhough the list of languages, and clicking each one individually. Each language have their own id, so you can select then individually!

Ah ok.... then it is probably something else!

**Comment by jpelay** [comment]
I'm still kinda blocked here, but found out something. This is how the logs look after printing `g.lang` in each `setup_language`. Here the first language, ar, have already been selected and processed and we're onto the next language, bg, the section with ### is in `change_language` so the language is being changed, and the section with *** is in each `setup_language`:


```
####################################################################################################
bg
####################################################################################################
[2023-03-14 11:45:05,153] app.py teardown_request_finish_logging 320 <DEBUG>: {'path': '/change_language', 'method': 'POST', 'remote_ip': '127.0.0.1', 'user_agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/12.0.2 Chrome/106.0.5249.51 Electron/21.0.0 Safari/537.36', 'start_time': '2023-03-14T15:45:05.152911Z', 'pid': 6373, 'loadavg': 0.73, 'fault': 0, 'http_code': 200, 'end_time': '2023-03-14T15:45:05.153467Z', 'user_ms': 0, 'sys_ms': 0, 'max_rss': 66756, 'inc_max_rss': 0, 'duration_ms': 0}
[2023-03-14 11:45:05,153] _internal.py _log 113 <INFO>: 127.0.0.1 - - [14/Mar/2023 11:45:05] "POST /change_language HTTP/1.1" 200 -
****************************************************************************************************
ar
****************************************************************************************************
[2023-03-14 11:45:05,194] app.py teardown_request_finish_logging 320 <DEBUG>: {'path': '/hedy', 'method': 'GET', 'remote_ip': '127.0.0.1', 'user_agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/12.0.2 Chrome/106.0.5249.51 Electron/21.0.0 Safari/537.36', 'start_time': '2023-03-14T15:45:05.175177Z', 'pid': 6373, 'loadavg': 0.73, 'fault': 0, 'load_yaml_pickled_ms': 0, 'load_yaml_pickled_cnt': 5, 'render_template_ms': 6, 'render_template_cnt': 1, 'http_code': 200, 'end_time': '2023-03-14T15:45:05.194598Z', 'user_ms': 19, 'sys_ms': 0, 'max_rss': 66864, 'inc_max_rss': 108, 'duration_ms': 19}
[2023-03-14 11:45:05,195] _internal.py _log 113 <INFO>: 127.0.0.1 - - [14/Mar/2023 11:45:05] "GET /hedy HTTP/1.1" 200 -
****************************************************************************************************
bg
****************************************************************************************************
[2023-03-14 11:45:05,819] app.py teardown_request_finish_logging 320 <DEBUG>: {'path': '/vendor/skulpt.min.js.map', 'method': 'GET', 'remote_ip': '127.0.0.1', 'user_agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/12.0.2 Chrome/106.0.5249.51 Electron/21.0.0 Safari/537.36', 'start_time': '2023-03-14T15:45:05.817109Z', 'pid': 6373, 'loadavg': 0.73, 'fault': 0, 'render_template_ms': 0, 'render_template_cnt': 1, 'http_code': 404, 'end_time': '2023-03-14T15:45:05.819514Z', 'user_ms': 2, 'sys_ms': 0, 'max_rss': 66896, 'inc_max_rss': 0, 'duration_ms': 2}
[2023-03-14 11:45:05,819] _internal.py _log 113 <INFO>: 127.0.0.1 - - [14/Mar/2023 11:45:05] "GET /vendor/skulpt.min.js.map HTTP/1.1" 404 -
****************************************************************************************************
bg
****************************************************************************************************
[2023-03-14 11:45:05,872] app.py teardown_request_finish_logging 320 <DEBUG>: {'path': '/js/appbundle.js.map', 'method': 'GET', 'remote_ip': '127.0.0.1', 'user_agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/12.0.2 Chrome/106.0.5249.51 Electron/21.0.0 Safari/537.36', 'start_time': '2023-03-14T15:45:05.871905Z', 'pid': 6373, 'loadavg': 0.73, 'fault': 0, 'http_code': 200, 'end_time': '2023-03-14T15:45:05.872597Z', 'user_ms': 0, 'sys_ms': 0, 'max_rss': 66896, 'inc_max_rss': 0, 'duration_ms': 0}
[2023-03-14 11:45:05,872] _internal.py _log 113 <INFO>: 127.0.0.1 - - [14/Mar/2023 11:45:05] "GET /js/appbundle.js.map HTTP/1.1" 200 -
****************************************************************************************************
bg
****************************************************************************************************
[2023-03-14 11:45:07,231] app.py teardown_request_finish_logging 320 <DEBUG>: {'path': '/parse', 'method': 'POST', 'remote_ip': '127.0.0.1', 'user_agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/12.0.2 Chrome/106.0.5249.51 Electron/21.0.0 Safari/537.36', 'start_time': '2023-03-14T15:45:06.103452Z', 'pid': 6373, 'loadavg': 0.73, 'fault': 0, 'level': 1, 'lang': 'ar', 'session_id': '72809ae8a04d4d18a8489bbcccfb6365', 'username': None, 'db_update:program-stats': 1, 'db_update_ms': 2, 'db_update_cnt': 1, 'transpile_ms': 508, 'transpile_cnt': 1, 'server_error': None, 'http_code': 200, 'end_time': '2023-03-14T15:45:07.231256Z', 'user_ms': 1126, 'sys_ms': 0, 'max_rss': 67408, 'inc_max_rss': 512, 'duration_ms': 1127}
[2023-03-14 11:45:07,231] _internal.py _log 113 <INFO>: 127.0.0.1 - - [14/Mar/2023 11:45:07] "POST /parse HTTP/1.1" 200 -
****************************************************************************************************
bg
****************************************************************************************************
[2023-03-14 11:45:07,306] app.py teardown_request_finish_logging 320 <DEBUG>: {'path': '/vendor/skulpt-stdlib-extensions.js', 'method': 'GET', 'remote_ip': '127.0.0.1', 'user_agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/12.0.2 Chrome/106.0.5249.51 Electron/21.0.0 Safari/537.36', 'start_time': '2023-03-14T15:45:07.305516Z', 'pid': 6373, 'loadavg': 0.73, 'fault': 0, 'http_code': 304, 'end_time': '2023-03-14T15:45:07.305984Z', 'user_ms': 0, 'sys_ms': 0, 'max_rss': 67408, 'inc_max_rss': 0, 'duration_ms': 0}
[2023-03-14 11:45:07,306] _internal.py _log 113 <INFO>: 127.0.0.1 - - [14/Mar/2023 11:45:07] "GET /vendor/skulpt-stdlib-extensions.js HTTP/1.1" 304 -
****************************************************************************************************
```

Notice how the only path that retains the language ar is in the path /hedy, and that is try for every language!

**Comment by jpelay** [comment]
Somehow is working now, after changing... nothing

image

**Comment by rix0rrr** [comment]
I changed the promise a little, maybe this is a cheat technique that might help!

image

There are three changes:

* First, we need to get a list of languages "from somewhere". For now, I hardcoded a list into the source. Potentially you could generate a JSON file with list of languages at build time or something, and have `cy.readFile()` read that list.
* Then, I **generate a test for every list in that language**. By doing a `for` loop, and calling `it('...', () => { .. })` *inside* the for loop, we get a new test for every language in the list.
* Finally, to make it easy to switch to a particular language without using the UI (cheating, I know), I added the ability to force the current language to a particular language by passing the query argument `?language=`. So if you add `?language=fr` to a URL, it will immediately cause a switch to French.

Does this help? Good luck with it!

**Comment by reedspool** [comment]
@rix0rrr below is not a response about your for loop strategy. I think you just nailed it your first time and no more changes are necessary but good to keep everyone aware of that sticky JS confusion!

I followed my (and Felienne's) suspicions about the for loop during the collaborator meeting. I just did a quick search `for loop cypress testing` and happened to find [this SO answer](https://stackoverflow.com/questions/52212868/cypress-io-writing-a-for-loop#comment96676166_53487016)  with a comment that confirms those suspicions.

Tl;dr: this problem is exactly what [`cy.each`](https://docs.cypress.io/api/commands/each) solves. See that answer and that those docs and if you still have questions about how to proceed I'd be happy to look further!

This is of course if we are not satisfied with @rix0rrr 's solid approach! That seems like a very reasonable tactic!

**Comment by rix0rrr** [comment]
@reedspool that SO post seems to be about using loops *inside* the test, vs what this code is doing which is creating multiple tests at module load time.

But it might be you are referring back to a discussion I wasn’t part of?

**Comment by jpelay** [comment]
Hi! @rix0rrr and @reedspool! 

@reedspool I quickly tried the SO approach about wrapping the array and running `each` but it still didn't work.
@rix0rrr  I tried yours and it ineed is a good start! I will work on this further.

Thank you both of you for the help!

**Comment by reedspool** [comment]
Glad you've got a good path forward @jpelay!

@rix0rrr Yes sorry about the confusion, I was continuing on thoughts we shared at the last Collaborators meeting.

**Comment by jpelay** [comment]
Hi all!! @Felienne I added the languages to test just hard coding there in the test. I will improve this once we tackle the test for the pygame features. But it now works! I edited `eywords-es.lark` and it now works.

image

**Comment by Felienne** [comment]
OMG what a tour de force this was!! So happy it is working now. I made an issue for the hard-coded list of languages, just so we don't forget: #4168
